plugins/woocommerce/src/Internal/ProductQueryFilters/FilterDataProvider.php

```diff
@@ -0,0 +1,364 @@
+<?php
+/**
+ * FilterDataProvider class file.
+ */
+
+namespace Automattic\WooCommerce\Internal\ProductQueryFilters;
+
+use WC_Cache_Helper;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Class for filter data.
+ */
+class FilterDataProvider {
+	/**
+	 * Instance of FilterClausesGenerator.
+	 *
+	 * @var FilterClausesGenerator
+	 */
+	private $filter_clauses_generator;
+
+	/**
+	 * Initialize dependencies.
+	 *
+	 * @internal
+	 *
+	 * @param FilterClausesGenerator $filter_clauses_generator Instance of FilterClausesGenerator.
+	 *
+	 * @return void
+	 */
+	final public function init( FilterClausesGenerator $filter_clauses_generator ): void {
+		$this->set_filter_clauses_generator( $filter_clauses_generator );
+	}
+
+	/**
+	 * Allow setting the clauses provider at run time.
+	 *
+	 * @param ClausesGeneratorInterface $filter_clauses_generator Instance of FilterClausesGenerator.
+	 *
+	 * @return void
+	 */
+	final public function set_filter_clauses_generator( ClausesGeneratorInterface $filter_clauses_generator ): void {
+		$this->filter_clauses_generator = $filter_clauses_generator;
+	}
+
+	/**
+	 * Get price data for current products.
+	 *
+	 * @param array $query_vars The WP_Query arguments.
+	 * @return object
+	 */
+	public function get_filtered_price( $query_vars ) {
```

Comment:
The methods `get_filtered_price`, `get_stock_status_counts`, `get_rating_counts`, and `get_attribute_counts` share a lot of structural similarities. WDY about reducing repetition by abstracting common logic into private methods, so we can stick to the DRY (Don't Repeat Yourself) principle?

Reply:
Yep, I agree. I planned (see https://github.com/woocommerce/woocommerce/pull/43772#issuecomment-1940744953) to address that in a follow-up PR that caches the enclosing query as @vedanshujain suggested in https://github.com/woocommerce/woocommerce/pull/43772#issuecomment-1935846563.